src/sentry/services/hybrid_cloud/user.py

```diff
@@ -0,0 +1,82 @@
+from abc import abstractmethod
+from typing import Iterable, List, Optional, Sequence
+
+from sentry.models.group import Group
+from sentry.models.organizationmember import OrganizationMember
+from sentry.models.user import User
+from sentry.services.hybrid_cloud import (
+    CreateStubFromBase,
+    InterfaceWithLifecycle,
+    silo_mode_delegation,
+)
+from sentry.silo import SiloMode
+
+
+class UserService(InterfaceWithLifecycle):
+    @abstractmethod
+    def get_users_by_email(self, email: str, is_active=True) -> List[User]:
+        """
+        Return a list of active users with verified emails matching the parameter
+        :param email:
+        A case insensitive email to match
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def get_from_group(self, group: Group) -> List[User]:
+        """Get all users in all teams in a given Group's project."""
+        pass
+
+    @abstractmethod
+    def get_users(self, user_ids: Iterable[int], is_active=True) -> List[User]:
+        """
+        This method returns User objects given an iterable of IDs
+        :param user_ids:
+        A list of user IDs to fetch
+        :return:
+        """
+        pass
+
+    def get_user(self, user_id: int, is_active=True) -> Optional[User]:
+        users = self.get_users([user_id], is_active=is_active)
+        if len(users) > 0:
+            return users[0]
+        else:
+            return None
+
+
+class DatabaseBackedUserService(UserService):
+    def get_users_by_email(self, email: str) -> Sequence[User]:
+        return User.objects.filter(
+            emails__is_verified=True, is_active=True, emails__email__iexact=email
+        )
+
+    def get_from_group(self, group: Group) -> List[User]:
+        group_memberships = OrganizationMember.objects.filter(
+            organization=group.organization,
+            teams__in=group.project.teams.all(),
+        ).values_list("id", flat=True)
+
+        return user_service.get_users(set(group_memberships))
+
+    def get_users(self, user_ids: Iterable[int], is_active=True) -> List[User]:
+
+        query = User.objects.filter(id__in=user_ids)
+        if is_active is not None:
+            query = query.filter(is_active=is_active)
+        return query
```

Comment:
If we're going to have both the db and remote implementations return the same thing we shouldn't be returning querysets as it is unlikely that we would build a queryset implementation for the RPC/cache backend. Should we cast to `list()` here as a quick fix?

Reply:
I did this for now. After some discussion with @corps, we actually think it may make sense to return ORM models in some cases (particularly where we expect a read-through cache), but to make the services return totally distinct models per silo with corresponding distinct ORM relationships. This gives total clarity when dealing with each different model

Reply:
I mean, hmm.
One thing, that would be useful @mikejihbe is if you did, still, extract a dataclass *even if we implement that dataclass to look a lot like an ORM* (for instance, adding helper properties that access the relation, etc).

I think it's possible to do this without colouring the entire world, but it starts to enforce certain guarantees of access.  When we're using the `user_service` in other places in the code, we should be informed when the impact of pulling objects from the service might change which relationships and other access we have based on the silo.

Even if we're going to use 2 different ORM classes in the future, I do think there is value whilst laying these APIs out to actually atleast split the class out from the existing ORM one.  Would it be possible you think to return something for now that you either 1.  forcibly upgrade into real User orm objects or 2.  is duct typed minimally to represent the needs of your callsites?

Reply:
I'm strugging between realities of how useful it is to get the interface extracted before we know what our final typing is going to be (or what the final RPC / caching layer looks like) with also the sense that dividing the types and seeing the deeper impact has been useful broadly for me, too.  Did you have further thoughts @markstory ?

I don't want this PR or any PR to become huge, but I'm pondering if there is a compromise.

Reply:
I think my final answer is this:
In the end, I think we want to have a smooth ORM like interface for typical developers.  I think that interface, however, does not exist purely on the service layer object.  I think in the end another layer helps translate to an ORM object that is local to the region.  Ideally that layer is invoked closer to controllers and *not* inside common helpers.

The "big problem" I want to avoid that can be insidious is helper functions that are written using any monolith ORM object that may unfortunately only have test coverage in one test silo by the end, but still be invoked and fail in the other silo.  

I want developers to see and understand that we are "colouring" the app across silo bounds and that yes, there is atleast some substantial enough difference in access now that they should be thinking about.  I want that to be obvious and I want it to be testable without it being silent.

Reply:
What that means for this particular PR is that I would prefer the user service return dumb objects, and we explicitly upgrade them into ORM objects.  That way when we expand on the usage of the user service we are forced to see the impact and start considering what the next step is ensuring that our system is still performant and easy to use.

Reply:
Or maybe not.  The User model may benefit in the end from consistent interface.  Arrrgh.  I don't know.  🤷   I literally jsut keep going back and forth.

Reply:
I think focusing on the resulting workflow is important as we'll be using these interfaces and decisions for a while. I'm of two minds still as well.

Returning ORM entities from the services is convenient for us in the present and in future scenarios when data is 'local' to the silo operations are being done it. Folks will generally be comfortable with using ORM entities and doing things like fetching relations. Using ORM entities are problematic in the multi-region world because of how powerful and flexible they are. We simply won't be able to do `user.save()` in most application code in the future. Instead we'll have to go through the service abstractions to have consistent behavior.

Using Data transfer objects or dumb entities is attractive as we can make both the local and remote implementations of the service the same, but we need to re-implement all the logic that User ORM objects have.

I'm partial to using ORM entities. My thinking is that:

1. We're already patching object managers to raise errors on incorrect silo use.
2. Using ORM objects will let us disrupt other teams less.
3. We don't need to come up with a solid abstraction for entities as we're using the one that django provides.

One looming concern with ORM entities is how we handle relations. The ORM allows incremental lazy loading through related managers and we simply don't have that luxury with remote service implementations.

Reply:
I'm putting together a doc on this subject in greater detail where we can locate further details on this subject.

I definitely was in favor of ORM entities in the case of Region resources initially, but my opinion on that has been evolving.  I encouraged mike to try using data classes here but it doesn't have to be a final decision.

Can I give one doc where I articulate, much better than I have, a definitive reason for the service class not to expose ORM objects?  Then we can discuss and come to the conclusion together.